src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -29,6 +31,19 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     /// </summary>
     internal class CompletionHandler : IRequestHandler<LSP.CompletionParams, LSP.CompletionList?>
     {
+        /// <summary>
+        /// This sets the max list size we will return in response to a completion request.
+        /// If there are more than this many items, we will set the isIncomplete flag on the returned completion list.
+        /// </summary>
+        private const int MaxCompletionListSize = 1000;
+
+        /// <summary>
+        /// This id is set when we return an incomplete completion list.
+        /// When the client asks us for more items  using <see cref="LSP.CompletionTriggerKind.TriggerForIncompleteCompletions"/>
+        /// we use this result id to lookup the previously calculated list.
+        /// </summary>
+        private long? _lastIncompleteResultId;
+
```

Comment:
For each client we create a new CompletionHandler, see https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Completion/CompletionHandlerProvider.cs,36

So by being a non-static member inside the completion handler this is already per client

Reply:
nit: up to you, but consider if it might be cleaner to just use a normal long/int, with a sentinel value (like -1) to indicate that there are no current incomplete members.

Reply:
I'll stick with nullable here - for me it makes it explicit where I need to check for no incomplete members